glib-macros/src/clone.rs

```diff
@@ -1,952 +1,487 @@
+
+                break;
             } else {
-                "]"
+                return Err(syn::Error::new_spanned(
+                        attr,
+                        "Unsupported attribute: only `strong`, `weak`, `weak_allow_none`, `to_owned` and `default_return` are supported",
+                ));
             }
         }
-        Delimiter::None => "",
-    }
-}
 
-fn group_to_string(g: &Group) -> String {
-    format!(
-        "{}{}{}",
-        delimiter_to_string(g.delimiter(), true),
-        tokens_to_string(PeekableProcIter::from(g)),
-        delimiter_to_string(g.delimiter(), false),
-    )
-}
-
-#[doc(alias = "get_expr")]
-fn expr(parts: &mut PeekableProcIter) -> Result<String, TokenStream> {
-    let mut ret = String::new();
-    let mut total = 0;
-    let span = parts.current_span;
-    match parts.next() {
-        Some(TokenTree::Literal(l)) => ret.push_str(&l.to_string()),
-        Some(TokenTree::Ident(i)) => ret.push_str(&i.to_string()),
-        Some(TokenTree::Punct(p)) => match p.to_string().as_str() {
-            "[" | "{" | "(" => {
-                total += 1;
-            }
-            x => {
-                return Err(parts
-                    .generate_error(&format!("Unexpected token `{x}` after `@default-return`")))
-            }
-        },
-        Some(TokenTree::Group(g)) => return Ok(group_to_string(&g)),
-        None => {
-            return Err(
-                parts.generate_error_with_span("Unexpected end after `@default-return`", span)
-            )
-        }
-    };
-    loop {
-        match parts.peek() {
-            Some(TokenTree::Punct(p)) => {
-                let p_s = p.to_string();
-                if p_s == "{" || p_s == "(" || p_s == "[" || p_s == "<" {
-                    total += 1;
-                } else if p_s == "}" || p_s == ")" || p_s == "]" || p_s == ">" {
-                    total -= 1;
-                } else if p_s == "," && total == 0 {
-                    return Ok(ret);
+        // Following is a closure or async block
+        let expr = input.parse::<Expr>()?;
+        let body = match expr {
+            Expr::Async(async_) => {
+                if async_.capture.is_none() {
+                    return Err(syn::Error::new_spanned(
+                        async_,
+                        "Async blocks with the `move` keyword required",
+                    ));
                 }
-                ret.push_str(&p_s);
-            }
-            Some(TokenTree::Group(g)) => {
-                ret.push_str(&group_to_string(g));
+
+                ClosureOrAsync::Async(async_)
             }
-            Some(x) => {
-                if total == 0 && !ret.ends_with(':') {
-                    return Ok(ret);
+            Expr::Closure(closure) => {
+                if closure.capture.is_none() {
+                    return Err(syn::Error::new_spanned(
+                        closure,
+                        "Closures with the `move` keyword required",
```

Comment:
Same as above. You should maybe write a small common function to only write the message in one place (same for the wrong number of attributes above).

Reply:
Both messages are actually different though